components/scream/src/physics/p3/p3_functions_cloud_sed_impl.hpp

```diff
@@ -0,0 +1,124 @@
+#ifndef P3_FUNCTIONS_CLOUD_SED_IMPL_HPP
+#define P3_FUNCTIONS_CLOUD_SED_IMPL_HPP
+
+#include "p3_functions.hpp"
+
+namespace scream {
+namespace p3 {
+
+/*
+ * Implementation of p3 cloud sedimentation function. Clients should NOT #include
+ * this file, #include p3_functions.hpp instead.
+ */
+
+template <typename S, typename D>
+KOKKOS_FUNCTION
+void Functions<S,D>
+::cloud_sedimentation(
+    const uview_1d<const Spack>& qc_incld,
+    const uview_1d<const Spack>& rho,
+    const uview_1d<const Spack>& inv_rho,
+    const uview_1d<const Spack>& lcldm,
+    const uview_1d<const Spack>& acn,
+    const uview_1d<const Spack>& inv_dzq,
+    const view_dnu_table& dnu,
+    const MemberType& team,
+    const Workspace& workspace,
+    const Int& nk, const Int& ktop, const Int& kbot, const Int& kdir, const Scalar& dt, const Scalar& odt, const bool& log_predictNc,
+    const uview_1d<Spack>& qc,
+    const uview_1d<Spack>& nc,
+    const uview_1d<Spack>& nc_incld,
+    const uview_1d<Spack>& mu_c,
+    const uview_1d<Spack>& lamc,
+    const uview_1d<Spack>& qc_tend,
+    const uview_1d<Spack>& nc_tend,
+    Scalar& prt_liq)
+{
+  // Get temporary workspaces needed for the cloud-sed calculation
+  uview_1d<Spack> V_qc, V_nc, flux_qx, flux_nx;
+  workspace.template take_many<4>(
+    {"V_qc", "V_nc", "flux_qx", "flux_nx"},
+    {&V_qc, &V_nc, &flux_qx, &flux_nx});
+
+  view_1d_ptr_array<Spack, 2>
+    fluxes_ptr = {&flux_qx, &flux_nx},
+    vs_ptr     = {&V_qc, &V_nc},
+    qnr_ptr    = {&qc, &nc};
+
+  // find top, determine qxpresent
+  const auto sqc = scalarize(qc);
+  bool log_qxpresent;
+  const Int k_qxtop = find_top(team, sqc, C::QSMALL, kbot, ktop, kdir, log_qxpresent);
+
+  if (log_qxpresent) {
+    Scalar dt_left = dt;    // time remaining for sedi over full model (mp) time step
+    Scalar prt_accum = 0.0; // precip rate for individual category
+
+    // find bottom
+    Int k_qxbot = find_bottom(team, sqc, C::QSMALL, kbot, k_qxtop, kdir, log_qxpresent);
+
+    while (dt_left > 1.e-4) {
```

Comment:
Let's move this magic number that all `dt_left` loops use to the constants.

Reply:
Sure, what should I call it?

Reply:
`dt_left_tol`?